pkg/cloud/azure/services/virtualmachines/virtualmachines_stack.go

```diff
@@ -251,3 +260,86 @@ func (s *StackHubService) Delete(ctx context.Context, spec azure.Spec) error {
+		dataDiskName := azure.GenerateDataDiskName(vmSpec.Name, disk.NameSuffix)
+
+		if len(dataDiskName) > 80 {
+			return nil, apierrors.InvalidMachineConfiguration("failed to create Data Disk: %s for vm %s. "+
+				"The overall disk name name must not exceed 80 chars in length. Check your `nameSuffix`.",
+				dataDiskName, vmSpec.Name)
+		}
+
+		if matched := reg.MatchString(disk.NameSuffix); !matched {
+			return nil, apierrors.InvalidMachineConfiguration("failed to create Data Disk: %s for vm %s. "+
+				"The nameSuffix can only contain letters, numbers, "+
+				"underscores, periods or hyphens. Check your `nameSuffix`.",
+				dataDiskName, vmSpec.Name)
+		}
+
+		if disk.DiskSizeGB < 4 {
+			return nil, apierrors.InvalidMachineConfiguration("failed to create Data Disk: %s for vm %s. "+
+				"`diskSizeGB`: %d, is invalid, disk size must be greater or equal than 4.",
+				dataDiskName, vmSpec.Name, disk.DiskSizeGB)
+		}
+
+		if _, exists := seenDataDiskNames[disk.NameSuffix]; exists {
+			return nil, apierrors.InvalidMachineConfiguration("failed to create Data Disk: %s for vm %s. "+
+				"A Data Disk with `nameSuffix`: %s, already exists. `nameSuffix` must be unique.",
+				dataDiskName, vmSpec.Name, disk.NameSuffix)
+		}
+
+		if disk.ManagedDisk.StorageAccountType == machinev1.StorageAccountUltraSSDLRS {
+			return nil, apierrors.InvalidMachineConfiguration("failed to create Data Disk: %s for vm %s. "+
+				"`storageAccountType`: %s, is not supported for Data Disk in Azure Stack Hub.",
+				dataDiskName, vmSpec.Name, machinev1.StorageAccountUltraSSDLRS, machinev1.CachingTypeNone)
+		}
+
+		if disk.ManagedDisk.DiskEncryptionSet != nil {
+			return nil, apierrors.InvalidMachineConfiguration("failed to create Data Disk: %s for vm %s. "+
+				"Disk Encryption Set is not supported for Data Disk in Azure Stack Hub.",
+				dataDiskName, vmSpec.Name)
+		}
+
+		if disk.Lun < 0 || disk.Lun > 63 {
+			return nil, apierrors.InvalidMachineConfiguration("failed to create Data Disk: %s for vm %s. "+
+				"Invalid value `lun`: %d. `lun` cannot be lower than 0 or higher than 63.",
+				dataDiskName, vmSpec.Name, disk.Lun)
+		}
+
+		if _, exists := seenDataDiskLuns[disk.Lun]; exists {
+			return nil, apierrors.InvalidMachineConfiguration("failed to create Data Disk: %s for vm %s. "+
+				"A Data Disk with `lun`: %d, already exists. `lun` must be unique.",
+				dataDiskName, vmSpec.Name, disk.Lun)
+		}
+
+		seenDataDiskNames[disk.NameSuffix] = struct{}{}
+		seenDataDiskLuns[disk.Lun] = struct{}{}
+
+		dataDisks[i] = compute.DataDisk{
+			CreateOption: compute.DiskCreateOptionTypesEmpty,
+			DiskSizeGB:   to.Int32Ptr(disk.DiskSizeGB),
+			Lun:          to.Int32Ptr(disk.Lun),
+			Name:         to.StringPtr(dataDiskName),
+			Caching:      compute.CachingTypes(disk.CachingType),
+		}
+
+		dataDisks[i].ManagedDisk = &compute.ManagedDiskParameters{
+			StorageAccountType: compute.StorageAccountTypes(disk.ManagedDisk.StorageAccountType),
+		}
```

Comment:
Was there a specific reason to omit support for Disk Encryption Set here?

https://github.com/mdbooth/machine-api-provider-azure/blob/afd0f3f9fdf04a106ae0be1a05e19c54171e5c28/pkg/cloud/azure/services/virtualmachines/virtualmachines.go#L624-L629